src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,219 @@
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     let is_foo = mutex.lock().unwrap().foo();
+    ///     match is_foo {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar();
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub SIGNIFICANT_DROP_IN_SCRUTINEE,
+    nursery,
+    "warns when a temporary of a type with a drop with a significant side-effect might have a surprising lifetime"
+}
+
+declare_lint_pass!(SignificantDropInScrutinee => [SIGNIFICANT_DROP_IN_SCRUTINEE]);
+
+impl<'tcx> LateLintPass<'tcx> for SignificantDropInScrutinee {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if has_significant_drop_in_scrutinee(cx, expr) {
+            span_lint_and_help(
+                cx,
+                SIGNIFICANT_DROP_IN_SCRUTINEE,
+                expr.span,
+                "The lifetimes of temporaries in the match scrutinee will last until the end of \
+                 the match expression, but one or more of them has a drop with a significant side-effect.",
+                None,
+                "The lifetimes of temporaries in a match scrutinee will last until the end of \
+                the match expression, but one of the temporaries in this match scrutinee has a drop \
+                with a significant side-effect (such as unlocking a mutex). Consider calling the \
+                function(s) outside of the match expression and using the result in the scrutinee.",
+            );
+        }
+    }
+}
+
+/// If the expression is an ExprKind::Match, check if the scrutinee has a significant drop that may
+/// have a surprising lifetime.
+fn has_significant_drop_in_scrutinee<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> bool {
+    match expr.kind {
+        ExprKind::Match(match_expr, _, _source) => has_significant_drop(cx, &match_expr, true),
+        _ => false,
+    }
+}
+
+/// Recursively check the expressions in the scrutinee to see if any of them have a return type that
+/// has `has_significant_drop` attribute. If one is found and it *is* the last method call, further
+/// check if there are previous method calls to determine if an arm is borrowing or taking ownership
+/// of the returned type. If there are subsequent method calls, we can infer that the lifetime of
+/// the returned type is *potentially* surprising, so return true. In all other cases, return false.
+fn has_significant_drop<'a, 'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'a>, is_last_in_chain: bool) -> bool {
+    if is_last_in_chain || !has_significant_drop_attribute(cx, cx.typeck_results().expr_ty(expr)) {
```

Comment:
I still think a visitor would be better here. The visitor impl would then just look like this:

```rust
// kinda pseudo code, please use better names and the like.
impl Visitor for V {
    fn visit_expr(.., expr) {
        match expr.peel_drop_temps().kind { // not sure if `peel_drop_temps` is necessary here
            ExprKind::MethodCall(_, exprs, ..) => if has_significant_drop_attr(cx, expr_ty(e)) { 
                self.sig_drop_found = true; 
                return; 
            } else { self.visit_expr(exprs[0]) }, // or call `walk_expr` depending on if you want to check all arguments
            ExprKind::<ShouldBeChecked> => same_as_method_call_case(),
            ExprKind::<ShouldBeSkipped> => {}, // like the `If` expression
            _ => walk_expr(expr), // Otherwise, just walk down the expression, this replaces all the cases where you just call has_significant_drop on all subexpressions.
        }
    }
}
```

Only checking the receiver of a `MethodCall` expression is a good way to implement the MVP. This can later be extended with other expression kinds that lead to temporaries that should be placed in front of the match.

Reply:
Thank you for the explanation and suggestion to look at intravisit! This makes a lot more sense now and I think I can make some good progress on it!